### PR TITLE
Add test coverage for AttributeCollection, fix bugs and nullable annotate

### PR DESCRIPTION
**Pull request by hughbe**
Add more test coverage and fix bugs/oversights associated with it

### CODE REVIEW

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs [hunk 1]
```diff
@@ -244,8 +256,13 @@ public bool Contains(Attribute[] attributes)
         /// Returns the default value for an attribute. This uses the following heuristic:
         /// 1. It looks for a public static field named "Default".
         /// </summary>
-        protected Attribute GetDefaultAttribute(Type attributeType)
+        protected Attribute? GetDefaultAttribute(Type attributeType)
         {
+            if (attributeType == null)
+            {
+                throw new ArgumentNullException(nameof(attributeType));
```
Comment (hughbe) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:263]: previously threw `ANE` with parameter name `key`

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs [hunk 2]
```diff
@@ -134,10 +136,15 @@ public static AttributeCollection FromExisting(AttributeCollection existing, par
...
         /// Gets the attribute with the specified type.
         /// </summary>
-        public virtual Attribute this[Type attributeType]
+        public virtual Attribute? this[Type attributeType]
         {
             get
             {
+                if (attributeType == null)
+                {
+                    throw new ArgumentNullException(nameof(attributeType));
```
Comment (hughbe) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:145]: previously threw `IndexOutOfRangeException` as we found an index in `_foundAttributeTypes` (which was initialised to all `null`) but this was out of the range of the empty collection.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs [hunk 3]
```diff
@@ -299,8 +316,13 @@ protected Attribute GetDefaultAttribute(Type attributeType)
         /// Determines if a specified attribute is the same as an attribute
         /// in the collection.
         /// </summary>
-        public bool Matches(Attribute attribute)
+        public bool Matches(Attribute? attribute)
         {
+            if (attribute == null)
```
Comment (hughbe) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:321]: I decided to be consistent with `Contains`. Let me know what you think

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs [hunk 4]
```diff
@@ -330,7 +357,7 @@ public bool Matches(Attribute[] attributes)
 
         bool ICollection.IsSynchronized => false;
 
-        object ICollection.SyncRoot => null;
+        object ICollection.SyncRoot => this;
```
Comment (hughbe) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:360]: should not be null

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs [hunk 5]
```diff
@@ -59,7 +61,7 @@ protected AttributeCollection() : this(Array.Empty<Attribute>())
         /// <summary>
         /// Creates a new AttributeCollection from an existing AttributeCollection
         /// </summary>
-        public static AttributeCollection FromExisting(AttributeCollection existing, params Attribute[] newAttributes)
+        public static AttributeCollection FromExisting(AttributeCollection existing, params Attribute[]? newAttributes)
```
Comment (stephentoub) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:64]: It's strange to see a params argument be nullable.  Do we actually expect folks to pass in a null array here, even though technically the implementation allows it?  I'd prefer to just remove this '?'.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs [hunk 6]
```diff
@@ -134,10 +136,15 @@ public static AttributeCollection FromExisting(AttributeCollection existing, par
         /// <summary>
         /// Gets the attribute with the specified type.
         /// </summary>
-        public virtual Attribute this[Type attributeType]
+        public virtual Attribute? this[Type attributeType]
```
Comment (stephentoub) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:139]: In what situation will this return null?

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs [hunk 7]
```diff
@@ -212,17 +219,22 @@ public static AttributeCollection FromExisting(AttributeCollection existing, par
         /// <summary>
         /// Determines if this collection of attributes has the specified attribute.
         /// </summary>
-        public bool Contains(Attribute attribute)
+        public bool Contains(Attribute? attribute)
```
Comment (stephentoub) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:222]: Why did you change this to allow null?  I'd prefer to just leave this method the way it was.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:321]: I don't think we should change this.  To my knowledge we've received 0 complaints about null not being allowed.  I'd just revert your changes to this method.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs [hunk 8]
```diff
@@ -315,8 +337,13 @@ public bool Matches(Attribute attribute)
         /// Determines if the attributes in the specified array are
         /// the same as the attributes in the collection.
         /// </summary>
-        public bool Matches(Attribute[] attributes)
+        public bool Matches(Attribute[]? attributes)
```
Comment (stephentoub) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:340]: Ditto

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs (hunk 6 above)
Comment (stephentoub) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:139]: I see that GetDefaultAttribute may return null.  Ok.

**Comment by hughbe** [comment]
Well this is odd. In the CI, the tests crash!
```
Executed on dci-mac-build-014.local
+ ./RunTests.sh --runtime-path /tmp/helix/working/ACD8097D/p
----- start Fri May 8 09:45:26 PDT 2020 =============== To repro directly: =====================================================
pushd .
/tmp/helix/working/ACD8097D/p/dotnet exec --runtimeconfig System.ComponentModel.TypeConverter.Tests.runtimeconfig.json --depsfile System.ComponentModel.TypeConverter.Tests.deps.json xunit.console.dll System.ComponentModel.TypeConverter.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing 
popd
===========================================================================================================
/private/tmp/helix/working/ACD8097D/w/A5640979/e /private/tmp/helix/working/ACD8097D/w/A5640979/e
  Discovering: System.ComponentModel.TypeConverter.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.ComponentModel.TypeConverter.Tests (found 1392 of 1393 test cases)
  Starting:    System.ComponentModel.TypeConverter.Tests (parallel test collections = on, max threads = 4)
./RunTests.sh: line 161: 66813 Bus error: 10           (core dumped) "$RUNTIME_PATH/dotnet" exec --runtimeconfig System.ComponentModel.TypeConverter.Tests.runtimeconfig.json --depsfile System.ComponentModel.TypeConverter.Tests.deps.json xunit.console.dll System.ComponentModel.TypeConverter.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing $RSP_FILE
/private/tmp/helix/working/ACD8097D/w/A5640979/e
----- end Fri May 8 09:46:51 PDT 2020 ----- exit code 138 ----------------------------------------------------------
```

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs [hunk 9]
```diff
@@ -1,6 +1,8 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+#nullable enable
```
Comment (safern) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:4]: Should we just annotate the whole assembly? We usually don't just annotate single files. @stephentoub what do you think?

Comment (hughbe) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:4]: I didn't really know how to do `TypeConverter` subclasses so left that unannotated

Comment (stephentoub) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:4]: I'd prefer such nullable annotations to be done in a dedicated PR and covering more of the assembly in one go.  Nullable annotations require a very specific kind of review, where you review not only what changed, but also what didn't change, since method signatures involving reference types implicitly change meanings when switching into a nullable context.

Comment (hughbe) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs]: I reverted the nullable annotation changes

File: src/libraries/System.ComponentModel.TypeConverter/tests/AttributeCollectionTests.cs [hunk 10]
```diff
@@ -13,183 +13,358 @@ public class AttributeCollectionTests
...
+            var collection = new SubAttributeCollection();
+            Assert.Null(collection.GetDefaultAttribute(attributeType));
 
-            Assert.Equal(count, attributeCollection.Count);
-            Assert.Equal(count, ((ICollection)attributeCollection).Count);
+            // Call again.
+            Assert.Null(collection.GetDefaultAttribute(attributeType));
+        }
+
+        private class AttributeWithDefaultField : Attribute
```
Comment (safern) [inline src/libraries/System.ComponentModel.TypeConverter/tests/AttributeCollectionTests.cs:212]: Maybe move these classes at the end of the file with the other test classes?

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs [hunk 11]
```diff
@@ -137,6 +137,11 @@ public static AttributeCollection FromExisting(AttributeCollection existing, par
         {
             get
             {
+                if (attributeType == null)
```
Comment (safern) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:140]: This is a virtual so once we nullable annotate this, all the overrides won't be able to accept null. It seems like they endup throwing because they try to access a dictionary with a null key, but just want to make sure we consider this. 

Also, we would need to update the docs: https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.attributecollection.item?view=netcore-3.1

**Review by safern — APPROVED** [review]
Modulo a couple comments

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs (hunk 11 above)
Comment (hughbe) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:140]: Yes this is deliberate. Not sure if you saw the comment but basically if `attributeType` was null, then we'd actually get a non-zero index, as the code is:
```cs
// 2 passes here for perf. Really!  first pass, we just
// check equality, and if we don't find it, then we
// do the IsAssignableFrom dance. Turns out that's
// a relatively expensive call and we try to avoid it
// since we rarely encounter derived attribute types
// and this list is usually short.
if (_foundAttributeTypes == null)
{
    _foundAttributeTypes = new AttributeEntry[FoundTypesLimit];
}

int ind = 0;

for (; ind < FoundTypesLimit; ind++)
{
    if (_foundAttributeTypes[ind].type == attributeType)
    {
        int index = _foundAttributeTypes[ind].index;
        if (index != -1)
        {
            return Attributes[index];
        }
        else
        {
            return GetDefaultAttribute(attributeType);
        }
    }
    if (_foundAttributeTypes[ind].type == null)
        break;
}
```

`_foundAttributeTypes[ind].type == attributeType)` would be true because everything is initialized internally to `null`. So therefore, when we do `return Attributes[index];` this throws `IndexOutOfRangeException` because the `_foundAttributeTypes` contains more elements (initialized to null) than `Attributes.Count`

Comment (hughbe) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:140]: See https://github.com/dotnet/runtime/pull/36118#discussion_r422224661

We could consider just using `List<Attribute>` but I didn't want to make these sorts of changes here

**Comment by danmosemsft** [comment]
@safern looks like @hughbe responded to your feedback. Is this good to merge?

**Comment by danmosemsft** [comment]
@safern any more feedback or can we merge this one?

**Comment by safern** [comment]
Looks good. Re-running CI as the OSX issue was fixed.

**Comment by danmosemsft** [comment]
Now CoreCLR Musl is failing https://dev.azure.com/dnceng/public/_build/results?buildId=848621&view=logs&j=73b7db51-6fbc-594d-25c9-15b2737b0bf6&t=c3ecdc4e-0ed6-5e18-3aef-dc02b04b6e41&l=10

**Comment by safern** [comment]
That is the previous build, but since the new one hasn't started GH hasn't overriden the badge because the HEAD of the branch is the same sha as for the previous run. The new build is this one:

https://dev.azure.com/dnceng/public/_build/results?buildId=849250&view=logs&j=e117a189-b428-527f-1281-65e8aef2c80d
